Operators Fixes (#44)

tajmone opened the pull request:
Move all operators definitions to independently named contexts, so that
they might included according to context in the future.

Fix comparison operators RegEx to ensure that bi-character operators
are captured as a single token, so that ligatures are shown correctly.

Amend capturing scope of logical operators from `keyword.logical` to
`keyword.operator.word`.

Ensure that all operators RegExs include `{{following_pointer}}` in their
definition, and add coverage tests for pointers following operators.

Update all operators tests accordingly.

tajmone committed:
Operators Fixes (#44)

Move all operators definitions to independently named contexts, so that
they might included according to context in the future.

Fix comparison operators RegEx to ensure that bi-character operators
are captured as a single token, so that ligatures are shown correctly.

Amend capturing scope of logical operators from `keyword.logical` to
`keyword.operator.word`.

Ensure that all operators RegExs include `{{following_pointer}}` in their
definition, and add coverage tests for pointers following operators.

Update all operators tests accordingly.

tajmone commented:
The build was failing due:

    /home/runner/work/purebasic-language-for-sublime-text/purebasic-language-for-sublime-text/Data/Packages/PureBasic/Tests/syntax_test_comparison_operators.pb:1: unable to find SYNTAX TEST marker

I accidentally changed the letter casing of one char in the first line, but didn't notice it since the tests worked locally. Now I've amended and force pushed, so it should be fine now.

But even if I've force-pushed the amend it doesn't seem that the build is testing it again (Travis CI would re-run the tests when force pushing), so I have no idea how to run the tests again.

peterthomashorn commented on Tests/syntax_test_comparison_operators.pb:
```diff
@@ -0,0 +1,41 @@
+﻿;; SYNTAX TEST "Packages/Purebasic/Purebasic.sublime-syntax"
```
I checked out the branch and this whole test file fails to load the referenced syntax definition because of the wrong capitalization you mentioned in the related issue. When this is correct all assertions are confirmed successfully.

tajmone replied:
I've already fixed that, but the GHAction doesn't restart when I force push, so it's not detecting the fix.

tajmone replied:
> You retained the .purebasic suffix in the syntax definition but dropped it in the assertions. Was that on purpose and if so, then why?

These are the best practices guidelines from the official community guide on syntaxes and syntax tests. Beside behind redundant, since the scope being tested is always only `source.purbasic`, it would also break tests in case in the future we add some further scope specifiers (e.g. `operator.comparison.gt`, etc.).

peterthomashorn replied:
I checked out your branch directly from your fork and it was not fixed.

According to [the official documentation](https://www.sublimetext.com/docs/scope_naming.html#syntax_definitions) the syntax name is recommended to be the last part of scope names. Hence I did it that way. Also, because all examples I have seen so far reflect that. On the other hand: you are right. They can be omitted.

tajmone replied:
> I checked out your branch directly from your fork and it was not fixed.

It was fixed on the PR branch as soon as I realized the problem (locally it runs fins, since Win OS has a case-insensitive files system), as you can see from the [changed files diff].

But the GitHub Actions Workflow is still reporting it in wrong casing.
If you look at the error reported at [line 12]:

```
/home/runner/work/purebasic-language-for-sublime-text/purebasic-language-for-sublime-text/Data/Packages/PureBasic/Tests/syntax_test_comparison_operators.pb:1 Error loading syntax file "Packages/Purebasic/Purebasic.sublime-syntax"
```

It still mentions the old letter casing mistake `Purebasic.sublime-syntax`.

I don't know why the job doesn't restart automatically when it detects a forced-push, but the error log is outdated. Either it lacks refresh settings, or it has wrong cache settings.

Can't you restart it manually?

<!----------------------------- REFERENCE LINKS ------------------------------>

[changed files diff]: https://github.com/peterthomashorn/purebasic-language-for-sublime-text/pull/47/files#diff-e82e01474ab0fd39837befc7345d9ec959d96fe776614b08f596a49c76de234dR1

[line 12]: https://github.com/peterthomashorn/purebasic-language-for-sublime-text/runs/4959927673?check_suite_focus=true#step:6:12

<!-- EOF -->

tajmone replied:
> There are options for case-sensitive file systems

Yes, now Windows 10 and above also have that option, on a per folder basis, but requires activating WSL. I haven't found the time to test it, but it's indeed a good feature for working with cross-platform repositories (Git or otherwise).

But I'm not sure if it can be enabled in system folders, including the ST packages folder which in Win OS is under `AppData/` — these sys folders might either be excluded from the feature, or unwise to fiddle with.

> I am not surprised the CI runner fails, I guess it uses ext4 or something else.

Surely, the CI will use a Unix like file system, but the problem is that the workflow doesn't restart on force pushes. You should look into it and find out why. On Travis CI, builds are automatically re-run when the PR branch changes (force pushes or otherwise); this is the default behaviour, which can be tweaked in the options (along with caching, etc.). The problem with these CI systems is that you end up reading docs for hours, and they also change quite often.

peterthomashorn commented on PureBasic.sublime-syntax:
```diff
@@ -415,3 +394,35 @@ contexts:
+    - include: operators_assignment
+    - include: operators_comparison
+    - include: operators_logical
+    - include: operators_bitwise
```
I used to insert a blank line between top level contexts for improved readability. If you are amending your branch because of the broken syntax definition reference, can you add this, too?

peterthomashorn replied:
This comment applies to the other operator contexts as well.

tajmone replied:
> I used to insert a blank line between top level contexts for improved readability. If you are amending your branch because of the broken syntax definition reference, can you add this, too?

I removed them because I though they might be the cause why the **Context Backtrace** is not working with the syntax — but now I think that backtrace functionality simply stopped working a few dev-updates ago, i.e. that it's an experimental features that is now broken, since I noticed that it stopped working with the native syntaxes too.

I'll try to restore the deleted empty lines, if I can't recover them for some reason I'll close the PR though.

tajmone replied:
Original empty lines restored. But again, after force pushing the workflow build is not triggered again, so you'll have to trigger it manually.

peterthomashorn commented on PureBasic.sublime-syntax:
```diff
@@ -415,3 +394,35 @@ contexts:
         0: variable.other.constant.purebasic
         1: punctuation.definition.variable.purebasic
+
+  operators_all:
```
I broke the pattern myself with the `constants` context as I see now. In case of an amendment, can you move this top level context and `constants` to the alphabetically correct position in the context list?

peterthomashorn commented on Tests/syntax_test_bitwise_operators.pb:
```diff
@@ -1,17 +1,36 @@
+;;                    ^     punctuation.definition.variable
+;;              ^^^         variable.other.constant
+;;              ^           punctuation.definition.variable
+;;                  ^       keyword.operator.bitwise
```
Isn't it better readable to list assertions in order of appearance when reading from left to right?

peterthomashorn commented on Tests/syntax_test_logical_operators.pb:
```diff
@@ -1,26 +1,52 @@
+If Left XOr Right : Debug #True : EndIf
+;;      ^^^   keyword.operator.word
+
+; Test for case-sensitivity
```
Don't these have the exact same capitalization? Did you mean to write them in lower case?

tajmone replied:
Yes I did intend to, but must have forgot.

tajmone replied:
Fixed

p4t5h3 reviewed (changes requested):
You retained the `.purebasic` suffix in the syntax definition but dropped it in the assertions. Was that on purpose and if so, then why?

p4t5h3 committed:
Affects #44 and #47: Requested changes.

- Fixed capitalization on a syntax definition file reference in a test.
- Reordered some syntax definition contexts alphabetically.
- Ordered some assertions in order of occurrence in the test line above.

peterthomashorn commented:
I am surprised I am allowed to push in a branch in our repository. Also I am used to the strict rule: reviewers do not change code but only request changes. But given the context I consider it acceptable to not be too strict here. 😊 This resolved the failing checks. In one test file the capitalization was still wrong.
